README.md
# go-opensea

[![Go Reference](https://pkg.go.dev/badge/github.com/mager/go-opensea.svg)](https://pkg.go.dev/github.com/mager/go-opensea)

A Go client for the OpenSea API.

## Endpoints supported

### v1

_https://docs.opensea.io/reference/api-overview_

- [x] Retrieving assets (GetAssets)
- [ ] Retrieving events (GetEvents)
- [ ] Retrieving collections (GetCollections)
- [ ] Retrieving bundles (GetBundles)
- [ ] Retrieving a single asset (GetAsset)
- [ ] Retrieving a single contract (GetContract)
- [x] Retrieving a single collection (GetCollection)
- [ ] Retrieving collection stats (GetCollectionStats)


## Example usage

```go
package main

import (
	"github.com/mager/go-opensea/opensea"
)

func main() {
	// Create a new client
	client := opensea.NewOpenSeaClient("")

	// Get all assets
	addy := "0x3b417FaeE9d2ff636701100891DC2755b5321Cc3" // Jay-Z's wallet address
	assets, err := client.GetAssets(addy)
	if err != nil {
		// TODO: Handle error
	}

	// Print the assets
	for _, asset := range assets {
		client.Log.Info(asset.Name)
	}
}
```



CHANGELOG.md
# CHANGELOG

## v0.2.0 - GetCollection

## v0.1.4 - GetAssets

## v0.1.0 - Init

opensea/asset_test.go
package opensea

import (
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"reflect"
	"testing"
	"time"

	"go.uber.org/zap"
	"go.uber.org/zap/zaptest"
)

func TestOpenSeaClient_GetAssetsWithOffset(t *testing.T) {
	type fields struct {
		Log          *zap.SugaredLogger
		apiKey       string
		client       *http.Client
		baseURL      string
		limitAssets  int
		requestDelay time.Duration
	}
	type args struct {
		owner  string
		offset int
	}
	tests := []struct {
		name        string
		fields      fields
		args        args
		path        string
		fixturePath string
		want        []Asset
		wantErr     bool
	}{
		{
			name: "Get assets with offset",
			fields: fields{
				Log:          zaptest.NewLogger(t).Sugar(),
				apiKey:       "",
				client:       &http.Client{},
				baseURL:      "https://api.opensea.io",
				limitAssets:  50,
				requestDelay: time.Millisecond * 250,
			},
			args: args{
				owner:  "0x3b417FaeE9d2ff636701100891DC2755b5321Cc3",
				offset: 0,
			},
			path:        "/api/v1/assets",
			fixturePath: "../testdata/get_assets.json",
			want:        FixtureGetAssetsResp,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				if r.URL.Path != tt.path {
					t.Errorf("Expected to request '%s', got: %s", tt.path, r.URL.Path)
				}
				if r.Header.Get("Content-Type") != "application/json" {
					t.Errorf("Expected Content-Type: application/json header, got: %s", r.Header.Get("Accept"))
				}
				w.WriteHeader(http.StatusOK)

				// Read the fixture
				jsonFile, err := os.Open(tt.fixturePath)
				if err != nil {
					t.Errorf("Failed to open fixture file: %s", err)
				}
				defer jsonFile.Close()

				// Write the fixture to the response
				jsonFile.Seek(0, 0)
				_, err = io.Copy(w, jsonFile)
				if err != nil {
					t.Errorf("Failed to write fixture to response: %s", err)
				}
			}))
			defer server.Close()

			c := &OpenSeaClient{
				Log:          tt.fields.Log,
				apiKey:       tt.fields.apiKey,
				client:       tt.fields.client,
				baseURL:      server.URL,
				limitAssets:  tt.fields.limitAssets,
				requestDelay: tt.fields.requestDelay,
			}
			got, err := c.GetAssetsWithOffset(tt.args.owner, tt.args.offset)
			if (err != nil) != tt.wantErr {
				t.Errorf("OpenSeaClient.GetAssetsWithOffset() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("OpenSeaClient.GetAssetsWithOffset() = %v, want %v", got, tt.want)
			}
		})
	}
}


opensea/asset.go
package opensea

import (
	"encoding/json"
	"fmt"
	"net/url"
	"time"
)

// Asset represents an asset on OpenSea.
// https://docs.opensea.io/reference/asset-object
type Asset struct {
	AnimationOriginalURL    string             `json:"animation_original_url"`
	AnimationURL            string             `json:"animation_url"`
	AssetContract           AssetAssetContract `json:"asset_contract"`
	BackgroundColor         string             `json:"background_color"`
	Collection              AssetCollection    `json:"collection"`
	Creator                 AssetCreator       `json:"creator"`
	Decimals                int                `json:"decimals"`
	Description             string             `json:"description"`
	ExternalLink            string             `json:"external_link"`
	ID                      int                `json:"id"`
	ImageOriginalURL        string             `json:"image_original_url"`
	ImagePreviewURL         string             `json:"image_preview_url"`
	ImageThumbnailURL       string             `json:"image_thumbnail_url"`
	ImageURL                string             `json:"image_url"`
	IsPresale               bool               `json:"is_presale"`
	LastSale                AssetLastSale      `json:"last_sale"`
	ListingDate             string             `json:"listing_date"`
	Name                    string             `json:"name"`
	NumSales                int                `json:"num_sales"`
	Owner                   AssetOwner         `json:"owner"`
	Permalink               string             `json:"permalink"`
	SellOrders              interface{}        `json:"sell_orders"`
	TokenID                 string             `json:"token_id"`
	TokenMetadata           string             `json:"token_metadata"`
	TopBid                  string             `json:"top_bid"`
	TransferFee             string             `json:"transfer_fee"`
	TransferFeePaymentToken string             `json:"transfer_fee_payment_token"`
	// TODO: Support traits
	// Traits                  []string           `json:"traits"`
}

type AssetAssetContract struct {
	Address                     string `json:"address"`
	AssetContractType           string `json:"asset_contract_type"`
	BuyerFeeBasisPoints         int    `json:"buyer_fee_basis_points"`
	CreatedDate                 string `json:"created_date"`
	DefaultToFiat               bool   `json:"default_to_fiat"`
	Description                 string `json:"description"`
	DevBuyerFeeBasisPoints      int    `json:"dev_buyer_fee_basis_points"`
	DevSellerFeeBasisPoints     int    `json:"dev_seller_fee_basis_points"`
	ExternalLink                string `json:"external_link"`
	ImageURL                    string `json:"image_url"`
	Name                        string `json:"name"`
	NftVersion                  string `json:"nft_version"`
	OnlyProxiedTransfers        bool   `json:"only_proxied_transfers"`
	OpenseaBuyerFeeBasisPoints  int    `json:"opensea_buyer_fee_basis_points"`
	OpenseaSellerFeeBasisPoints int    `json:"opensea_seller_fee_basis_points"`
	OpenseaVersion              string `json:"opensea_version"`
	Owner                       int    `json:"owner"`
	PayoutAddress               string `json:"payout_address"`
	SchemaName                  string `json:"schema_name"`
	SellerFeeBasisPoints        int    `json:"seller_fee_basis_points"`
	Symbol                      string `json:"symbol"`
	TotalSupply                 string `json:"total_supply"`
}

type AssetCollection struct {
	BannerImageURL              string           `json:"banner_image_url"`
	ChatURL                     string           `json:"chat_url"`
	CreatedDate                 string           `json:"created_date"`
	DefaultToFiat               bool             `json:"default_to_fiat"`
	Description                 string           `json:"description"`
	DevBuyerFeeBasisPoints      string           `json:"dev_buyer_fee_basis_points"`
	DevSellerFeeBasisPoints     string           `json:"dev_seller_fee_basis_points"`
	DiscordURL                  string           `json:"discord_url"`
	DisplayData                 AssetDisplayData `json:"display_data"`
	ExternalURL                 string           `json:"external_url"`
	Featured                    bool             `json:"featured"`
	FeaturedImageURL            string           `json:"featured_image_url"`
	Hidden                      bool             `json:"hidden"`
	ImageURL                    string           `json:"image_url"`
	InstagramUsername           string           `json:"instagram_username"`
	IsSubjectToWhitelist        bool             `json:"is_subject_to_whitelist"`
	LargeImageURL               string           `json:"large_image_url"`
	MediumUsername              string           `json:"medium_username"`
	Name                        string           `json:"name"`
	OnlyProxiedTransfers        bool             `json:"only_proxied_transfers"`
	OpenseaBuyerFeeBasisPoints  string           `json:"opensea_buyer_fee_basis_points"`
	OpenseaSellerFeeBasisPoints string           `json:"opensea_seller_fee_basis_points"`
	PayoutAddress               string           `json:"payout_address"`
	RequireEmail                bool             `json:"require_email"`
	SafelistRequestStatus       string           `json:"safelist_request_status"`
	ShortDescription            string           `json:"short_description"`
	Slug                        string           `json:"slug"`
	TelegramURL                 string           `json:"telegram_url"`
	TwitterUsername             string           `json:"twitter_username"`
	WikiURL                     string           `json:"wiki_url"`
}

type AssetOwner struct {
	Address       string    `json:"address"`
	Config        string    `json:"config"`
	ProfileImgURL string    `json:"profile_img_url"`
	User          AssetUser `json:"user"`
}

type AssetCreator struct {
	Address       string    `json:"address"`
	Config        string    `json:"config"`
	ProfileImgURL string    `json:"profile_img_url"`
	User          AssetUser `json:"user"`
}

type AssetUser struct {
	Username string `json:"username"`
}

type AssetDisplayData struct {
	CardDisplayStyle string `json:"card_display_style"`
}

type AssetLastSale struct {
	Asset AssetLastSaleAsset `json:"asset"`
}

type AssetLastSaleAsset struct {
	TokenID  string `json:"token_id"`
	Decimals int    `json:"decimals"`
}

type GetAssetsResponse struct {
	Assets []Asset `json:"assets"`
}

// GetAssetsWithOffset gets a list of assets with an offset
// https://docs.opensea.io/reference/getting-assets
func (c *OpenSeaClient) GetAssetsWithOffset(owner string, offset int) ([]Asset, error) {
	u, err := url.Parse(fmt.Sprintf("%s/api/v1/assets", c.baseURL))
	if err != nil {
		c.Log.Errorf("Error parsing url: %s", err)
		return nil, err
	}

	// Set query params
	q := u.Query()
	q.Set("owner", owner)
	q.Set("limit", fmt.Sprint(c.limitAssets))
	q.Set("offset", fmt.Sprint(offset))
	u.RawQuery = q.Encode()

	resp, err := c.Get(u)
	if err != nil {
		c.Log.Errorf("Error getting assets: %s", err)
		return nil, err
	}
	defer resp.Body.Close()

	var osResp GetAssetsResponse
	err = json.NewDecoder(resp.Body).Decode(&osResp)
	if err != nil {
		c.Log.Errorf("Error decoding response: %s", err)
		return nil, err
	}

	// TODO: Filter out assets with hidden collections
	// var filtered = []Asset{}
	// for _, asset := range osResp.Assets {
	// 	if !asset.Collection.Hidden {
	// 		filtered = append(filtered, asset)
	// 	}
	// }

	return osResp.Assets, nil
}

// GetAssets returns the assets for an address
func (c *OpenSeaClient) GetAssets(address string) ([]Asset, error) {
	var (
		allAssets []Asset
		offset    int
	)

	for {
		assets, err := c.GetAssetsWithOffset(address, offset)
		if err != nil {
			return allAssets, err
		}

		if len(assets) == 0 {
			break
		}

		allAssets = append(allAssets, assets...)
		offset += c.limitAssets
		time.Sleep(c.requestDelay)
	}

	return allAssets, nil
}


opensea/collection.go
package opensea

import (
	"encoding/json"
	"fmt"
	"net/url"
)

type Collection struct {
	Editors                     []string                          `json:"editors"`
	PaymentTokens               []CollectionPaymentTokens         `json:"payment_tokens"`
	PrimaryAssetContracts       []CollectionPrimaryAssetContracts `json:"primary_asset_contracts"`
	Stats                       CollectionStats                   `json:"stats"`
	BannerImageURL              string                            `json:"banner_image_url"`
	ChatURL                     string                            `json:"chat_url"`
	CreatedDate                 string                            `json:"created_date"`
	DefaultToFiat               bool                              `json:"default_to_fiat"`
	Description                 string                            `json:"description"`
	DevBuyerFeeBasisPoints      string                            `json:"dev_buyer_fee_basis_points"`
	DevSellerFeeBasisPoints     string                            `json:"dev_seller_fee_basis_points"`
	DiscordURL                  string                            `json:"discord_url"`
	DisplayData                 CollectionDisplayData             `json:"display_data"`
	ExternalURL                 string                            `json:"external_url"`
	Featured                    bool                              `json:"featured"`
	FeaturedImageURL            string                            `json:"featured_image_url"`
	Hidden                      bool                              `json:"hidden"`
	SafelistRequestStatus       string                            `json:"safelist_request_status"`
	ImageURL                    string                            `json:"image_url"`
	IsSubjectToWhitelist        bool                              `json:"is_subject_to_whitelist"`
	LargeImageURL               string                            `json:"large_image_url"`
	MediumUsername              string                            `json:"medium_username"`
	Name                        string                            `json:"name"`
	OnlyProxiedTransfers        bool                              `json:"only_proxied_transfers"`
	OpenseaBuyerFeeBasisPoints  string                            `json:"opensea_buyer_fee_basis_points"`
	OpenseaSellerFeeBasisPoints string                            `json:"opensea_seller_fee_basis_points"`
	PayoutAddress               string                            `json:"payout_address"`
	RequireEmail                bool                              `json:"require_email"`
	ShortDescription            string                            `json:"short_description"`
	Slug                        string                            `json:"slug"`
	TelegramURL                 string                            `json:"telegram_url"`
	TwitterUsername             string                            `json:"twitter_username"`
	InstagramUsername           string                            `json:"instagram_username"`
	WikiURL                     string                            `json:"wiki_url"`
}

type CollectionPaymentTokens struct {
	ID       int     `json:"id"`
	Symbol   string  `json:"symbol"`
	Address  string  `json:"address"`
	ImageURL string  `json:"image_url"`
	Name     string  `json:"name"`
	Decimals int     `json:"decimals"`
	EthPrice float64 `json:"eth_price"`
	UsdPrice float64 `json:"usd_price"`
}

type CollectionPrimaryAssetContracts struct {
	Address                     string `json:"address"`
	AssetContractType           string `json:"asset_contract_type"`
	CreatedDate                 string `json:"created_date"`
	Name                        string `json:"name"`
	NftVersion                  string `json:"nft_version"`
	OpenseaVersion              string `json:"opensea_version"`
	Owner                       int    `json:"owner"`
	SchemaName                  string `json:"schema_name"`
	Symbol                      string `json:"symbol"`
	TotalSupply                 string `json:"total_supply"`
	Description                 string `json:"description"`
	ExternalLink                string `json:"external_link"`
	ImageURL                    string `json:"image_url"`
	DefaultToFiat               bool   `json:"default_to_fiat"`
	DevBuyerFeeBasisPoints      int    `json:"dev_buyer_fee_basis_points"`
	DevSellerFeeBasisPoints     int    `json:"dev_seller_fee_basis_points"`
	OnlyProxiedTransfers        bool   `json:"only_proxied_transfers"`
	OpenseaBuyerFeeBasisPoints  int    `json:"opensea_buyer_fee_basis_points"`
	OpenseaSellerFeeBasisPoints int    `json:"opensea_seller_fee_basis_points"`
	BuyerFeeBasisPoints         int    `json:"buyer_fee_basis_points"`
	SellerFeeBasisPoints        int    `json:"seller_fee_basis_points"`
	PayoutAddress               string `json:"payout_address"`
}

type CollectionStats struct {
	OneDayVolume          float64 `json:"one_day_volume"`
	OneDayChange          float64 `json:"one_day_change"`
	OneDaySales           float64 `json:"one_day_sales"`
	OneDayAveragePrice    float64 `json:"one_day_average_price"`
	SevenDayVolume        float64 `json:"seven_day_volume"`
	SevenDayChange        float64 `json:"seven_day_change"`
	SevenDaySales         float64 `json:"seven_day_sales"`
	SevenDayAveragePrice  float64 `json:"seven_day_average_price"`
	ThirtyDayVolume       float64 `json:"thirty_day_volume"`
	ThirtyDayChange       float64 `json:"thirty_day_change"`
	ThirtyDaySales        float64 `json:"thirty_day_sales"`
	ThirtyDayAveragePrice float64 `json:"thirty_day_average_price"`
	TotalVolume           float64 `json:"total_volume"`
	TotalSales            float64 `json:"total_sales"`
	TotalSupply           float64 `json:"total_supply"`
	Count                 float64 `json:"count"`
	NumOwners             int     `json:"num_owners"`
	AveragePrice          float64 `json:"average_price"`
	NumReports            int     `json:"num_reports"`
	MarketCap             float64 `json:"market_cap"`
	FloorPrice            float64 `json:"floor_price"`
}

type CollectionDisplayData struct {
	CardDisplayStyle string `json:"card_display_style"`
}

type GetCollectionResponse struct {
	Collection Collection `json:"collection"`
}

// GetCollection gets a collection by slug
// https://docs.opensea.io/reference/retrieving-a-single-collection
func (c *OpenSeaClient) GetCollection(slug string) (Collection, error) {
	var collection Collection

	u, err := url.Parse(fmt.Sprintf("%s/api/v1/collection/%s", c.baseURL, slug))
	if err != nil {
		c.Log.Errorf("Error parsing url: %s", err)
		return collection, err
	}

	resp, err := c.Get(u)
	if err != nil {
		c.Log.Errorf("Error getting collection: %s", err)
		return collection, err
	}

	defer resp.Body.Close()

	var osResp GetCollectionResponse
	err = json.NewDecoder(resp.Body).Decode(&osResp)
	if err != nil {
		c.Log.Errorf("Error decoding response: %s", err)
		return collection, err
	}

	return osResp.Collection, nil
}


opensea/opensea.go
package opensea

import (
	"bytes"
	"encoding/json"
	"net/http"
	"net/url"
	"time"

	"go.uber.org/zap"
)

// Client represents the client for the OpenSea API.
type Client struct {
	apiKey      string
	client      *http.Client
	baseURL     *url.URL
	limitAssets int
	logger      *zap.SugaredLogger
	rateLimit   time.Duration
}

// NewClient creates a new OpenSea client with configuration.
func NewClient(apiKey string) *Client {
	return &Client{
		apiKey:  apiKey,
		baseURL: &url.URL{Scheme: "https", Host: "api.opensea.io", Path: "/"},
		client: &http.Client{
			Timeout: time.Second * 10,
		},
		limitAssets: 50,
		logger:      zap.NewExample().Sugar(),
		rateLimit:   time.Millisecond * 250,
	}
}

// NewRequest creates a new request and adds authentication headers.
func (c *Client) NewRequest(method string, u *url.URL, body interface{}) (*http.Request, error) {
	var (
		err error
		buf []byte
	)

	if body != nil {
		buf, err = json.Marshal(body)
		if err != nil {
			return nil, err
		}
	}

	req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(buf))
	if err != nil {
		return nil, err
	}

	req.Header.Set("X-API-KEY", c.apiKey)
	req.Header.Set("Content-Type", "application/json")

	return req, nil
}
